### PR TITLE
binding/c: autogen datatype api

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Auto generate datatype C bindings

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
@raffenet This PR is ready for your review.

**Review by raffenet — COMMENTED** [review]
Could we split off the last 3 commits into a new PR? All the code moving around stuff looks OK to me, save for my one comment about squashing. I'd like to review the autogen stuff on its own so I can look at the manpage generation more closely.

**Comment by hzhou** [comment]
> Could we split off the last 3 commits into a new PR? All the code moving around stuff looks OK to me, save for my one comment about squashing. I'd like to review the autogen stuff on its own so I can look at the manpage generation more closely.

Sure.

File: src/include/mpir_datatype.h [hunk 1]
```diff
@@ -558,6 +558,8 @@ void MPIR_Pack_size_impl(int incount, MPI_Datatype datatype, MPI_Aint * size);
 void MPIR_Type_lb_impl(MPI_Datatype datatype, MPI_Aint * displacement);
 
 /* Datatype functions */
+int MPIR_Type_create_resized(MPI_Datatype oldtype, MPI_Aint lb, MPI_Aint extent,
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_datatype.h]: The commits are split to #5009 with the above mentioned commit squashed (it was meant to be squashed but forgot). I'll rebase this PR once #5009 gets merged.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/binding/c/datatype_api.txt [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+    MPIR_Assert(size_x >= 0);
+    /* handle overflow: see MPI-3 p.104 */
+    *size = (size_x > INT_MAX) ? MPI_UNDEFINED : (int) size_x;
+}
+
+MPI_Type_size_x:
+    .desc: Return the number of bytes occupied by entries in the datatype
+    .skip: initcheck
+
+MPI_Type_set_name:
```
Comment (raffenet) [inline src/binding/c/datatype_api.txt:331]: missing `MPI_ERR_OTHER`

File: src/binding/c/datatype_api.txt [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    consider declaring the Fortran array with a zero origin
+    .vb
+        integer a(0:99)
+    .ve
+*/
+
+MPI_Type_create_resized:
+    .desc: Create a datatype with a new lower bound and extent from an existing datatype
+
+MPI_Type_create_struct:
```
Comment (raffenet) [inline src/binding/c/datatype_api.txt:405]: missing `MPI_ERR_TYPE`

File: src/binding/c/datatype_api.txt [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+    .desc: Create an MPI datatype from a general set of datatypes, displacements, and block sizes
+
+MPI_Type_dup:
+    .desc: Duplicate a datatype
+
+MPI_Type_hindexed:
```
Comment (raffenet) [inline src/binding/c/datatype_api.txt:411]: missing `MPI_ERR_TYPE, MPI_ERR_COUNT, MPI_ERR_ARG`

File: src/binding/c/datatype_api.txt [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+    it actually sends "a(1+1),a(1+11),...".
+    
+    If you wish to consider the displacements as array_of_displacements into a Fortran array,
+    consider declaring the Fortran array with a zero origin
+    .vb
+        integer a(0:99)
+    .ve
+*/
+
+MPI_Type_struct:
```
Comment (raffenet) [inline src/binding/c/datatype_api.txt:482]: missing `MPI_ERR_TYPE, MPI_ERR_COUNT`

File: src/binding/c/datatype_api.txt [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+    for the structure foo
+    .vb
+        blen[0] = 1; array_of_displacements[0] = 0; oldtypes[0] = MPI_INT;
+        blen[1] = 1; array_of_displacements[1] = &foo.b - &foo; oldtypes[1] = MPI_CHAR;
+        blen[2] = 1; array_of_displacements[2] = sizeof(foo); oldtypes[2] = MPI_UB;
+        MPI_Type_struct(3, blen, array_of_displacements, oldtypes, &newtype);
+    .ve
+*/
+
+MPI_Type_vector:
```
Comment (raffenet) [inline src/binding/c/datatype_api.txt:530]: missing `MPI_ERR_OTHER`

**Review by raffenet — COMMENTED** [review]
Found some differences. I'm not sure the places where `MPI_ERR_OTHER` is missing actually require adding it back, but I wanted to at least note it.

File: src/binding/c/datatype_api.txt [hunk 7]
```diff
@@ -0,0 +1,542 @@
+# vim: set ft=c:
+
+MPI_Address:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:3]: It's a removed function. The user should refer to the replacement function if they are reading this man page at all. My preference is not to bother maintaining the notes or specific error codes for deprecated/removed functions.

File: src/binding/c/datatype_api.txt [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+    .desc: Gets the address of a location in memory
+    .skip: global_cs
+    .replace: removed with MPI_Get_address
+
+MPI_Get_address:
+    .desc: Get the address of a location in memory
+    .skip: Fortran
+    .skip: global_cs
+/* -- notes-2 --
+    .N Fortran
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:15]: Fixed. I shouldn't put two separate `.skip:` lines. The latter overwrites the former.

File: src/binding/c/datatype_api.txt [hunk 9]
```diff
@@ -0,0 +1,542 @@
+# vim: set ft=c:
+
+MPI_Address:
+    location: BUFFER, asynchronous=True, suppress=f08_intent, [location in caller memory]
+    address: DISPLACEMENT, direction=out, [address of location]
+    .desc: Gets the address of a location in memory
+    .skip: global_cs
+    .replace: removed with MPI_Get_address
+
+MPI_Get_address:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:10]: Fixed.

File: src/binding/c/datatype_api.txt [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+*/
+{
+    MPI_Aint count_x;
+    MPIR_Get_count_impl(status, datatype, &count_x);
+
+    /* clip the value if it cannot be correctly returned to the user */
+    *count = (count_x > INT_MAX) ? MPI_UNDEFINED : (int) count_x;
+}
+
+MPI_Get_elements:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:67]: Fixed.

File: src/binding/c/datatype_api.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:3]: Many of our routines internally will return `MPI_ERR_OTHER`. We could always put that error code in our man pages just like `MPI_SUCCESS`. What do you think?

File: src/binding/c/datatype_api.txt [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+    .desc: Returns the number of basic elements
+    .skip: initcheck
+    .skip: global_cs
+{
+    MPI_Count byte_count = MPIR_STATUS_GET_COUNT(*status);
+    mpi_errno = MPIR_Get_elements_x_impl(&byte_count, datatype, count);
+    MPIR_ERR_CHECK(mpi_errno);
+}
+
+MPI_Pack:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:90]: > 
> 
> Also, the order of argument descriptions puts `Input/Output` arguments first. Is that how the manpage will render? Seems a bit odd.

Yes, the manpage renders the same order.  I believe that was the order when I checked the `pt2pt` original code. I can easily change the order. How about `Input`, then `Input/Output`, then `Output`?

File: src/binding/c/datatype_api.txt [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+    applies to both 'MPI_Pack' and 'MPI_Unpack'.  That is, the value 'size' is
+    the maximum that is needed by either 'MPI_Pack' or 'MPI_Unpack'.
+*/
+{
+    MPI_Aint size_x;
+    MPIR_Pack_size_impl(incount, datatype, &size_x);
+    MPIR_Assign_trunc(*size, size_x, int);
+}
+
+MPI_Unpack:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:142]: Hmm... the `outcount` is a `POLYDTYPE_NUM_ELEM` kind, which is the same kind as a `stride`, which can be negative. Thus there is validation on `POLYDTYPE_NUM_ELEM`. Typical count is of kind `POLYDTYPE_NUM_ELEM_NNI`.

Does the standard really allow unpacking negative number of elements here? I think it is probably an error in the standard's python binding.

Let me add a custom error checking here.

[the remark this answers is not in the document]
Comment (raffenet) [reply src/binding/c/datatype_api.txt:142]: > Does the standard really allow unpacking negative number of elements here? I think it is probably an error in the standard's python binding.

gosh I hope not 😄

File: src/binding/c/datatype_api.txt [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+    if (*datatype == MPI_FLOAT_INT || *datatype == MPI_DOUBLE_INT ||
+        *datatype == MPI_LONG_INT || *datatype == MPI_SHORT_INT ||
+        *datatype == MPI_LONG_DOUBLE_INT) {
+        goto fn_exit;
+    }
+}
+
+MPI_Type_free:
+    .desc: Frees the datatype
+{ -- early_return --
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:211]: Updated that into a `error_check` and missing notes added.

File: src/binding/c/datatype_api.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/binding/c/datatype_api.txt:90]: > Yes, the manpage renders the same order. I believe that was the order when I checked the `pt2pt` original code. I can easily change the order. How about `Input`, then `Input/Output`, then `Output`?

Sure, that works.

File: src/binding/c/datatype_api.txt (hunk 3 above)
Comment (hzhou) [reply src/binding/c/datatype_api.txt:405]: It is also missing the error checking for type arrays. I updated the python script to fix it.

File: src/binding/c/datatype_api.txt (hunk 4 above)
Comment (hzhou) [reply src/binding/c/datatype_api.txt:411]: Let's ignore the removed functions

File: src/binding/c/datatype_api.txt (hunk 5 above)
Comment (hzhou) [reply src/binding/c/datatype_api.txt:482]: Let's ignore the removed functions

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/binding/c/datatype_api.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:3]: Added commit to python script to always document `MPI_ERR_OTHER`.

File: src/binding/c/datatype_api.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/datatype_api.txt:103]: done.

File: src/binding/c/datatype_api.txt (hunk 2 above)
Comment (hzhou) [reply src/binding/c/datatype_api.txt:369]: Always added now.

File: src/binding/c/datatype_api.txt (hunk 6 above)
Comment (hzhou) [reply src/binding/c/datatype_api.txt:568]: Always added now.

**Comment by hzhou** [comment]
@raffenet Ready for your review again.

**Review by raffenet — APPROVED** [review]

**Pull request merged**